controller-server/src/main/java/com/yahoo/vespa/hosted/controller/versions/VespaVersion.java

```diff
@@ -136,4 +135,28 @@ public boolean equals(Object other) {
 
     }
 
+    private static boolean nonCanaryApplicationsBroken(ApplicationList failingOnThis,
+                                                       ApplicationList productionOnThis,
+                                                       Instant releasedAt) {
+        // all below are number for non-canary applications
+        int failing = failingOnThis.without(UpgradePolicy.canary).startedFailingAfter(releasedAt).size();
+        int production = productionOnThis.without(UpgradePolicy.canary).size();
+        int all = production + failing;
+
+        if (all == 0) return false;
+
+        double brokenByThisVersion = (double) failing / (double) all;
+        // limit for 'broken' depends on total number of applications on this version
+        double limit;
+        if (all <= 4) {
+            limit = 1.0;
+        } else if (all <= 10) {
+            limit = 0.75;
+        } else {
+            limit = 0.50;
+        }
```

Comment:
I think this is way too aggressive: If we have upgraded 100 applications and 50 fail, we should really not continue ... I think 10%, not 50% is about right when we have enough data.

Reply:
I agree with Jon. I suggest simplifying this by doing: if ( all <= 50 ) all=50; and otherwise only use the 10% limit. This means we immediately fail when we have 5 failing before we have upgraded 50 applications, which is the "enough data" criteria, then 10% after this. If needed we can tune the two parameters independently.

Reply:
You need to tolerate 1 of 1 failing, and 2 of 2 and so on up to a limit (4 of 4 is the limit now). But say you have 3 of 4 failing and number 5 is succesful. Then if 50% failure is accepted you will conclude it's broken and not continue, which doesn't sound like what we want. 

I made this mostly to have something to discuss with Martin, it's not finished yet. 

10% failures after 50 applications sounds good, though.

Reply:
How about 10% of 100 and then linearly down to 100% of 5:

```
applicationsGivingMaxConfidence = 100
applicationsGivingMinConfidence = 5
failureRatioAtMaxConfidence = 0.1

confidence = (all - applicationsGivingMinConfidence ) / 
                       (applicationsGivingMaxConfidence - applicationsGivingMinConfidence) 
if (confidence > 1.0) confidence = 1.0
if (confidence < 0.0) confidence = 0.0

failureToleranceRatio = 1 - confidence * (1 - failureRationAtMaxConfidence)  // called "limit" in your code
```

Reply:
Those 3 constants are easier to discuss imho, and we could tighten them as we reduce noise.

Btw, another thing we could do to reduce noise was to retry a bit for each application before concluding anything. It requires holding some more state though, so I would like to hear from @bjormel whether it actually helps much: Are many of the failures on the last upgrades due to instability which is fixed simply by retrying the same job?

Reply:
@hmusum I am not sure if your full comment related to my suggestion? My intention is to tolerate up to 5 failing deployments among the 50 first deployments, then 10% of all deployments after this.

@bratseth I think you suggestion is good; if you replace 100 with 50 I believe we would have the exact same handling. Your suggestion give more flexibility to the tuning, though.

Reply:
Btw; my biggest problem with reasoning about the initial suggestion is that it is discontinuous; e.g. 7 of 10 will not be seen as broken, while 6 of 11 will.